### PR TITLE
[DRAFT] Adds conditional-update support

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
- Adds support for conditional updates as described in the spec https://www.hl7.org/fhir/http.html#cond-update.
- Adds a Feature flag that can enable Conditional-Update conditionally in FHIR Server deployments
- Sets the `conditionalUpdate` capability in the capability statement

Limitations:
- Several race conditions have been identified where a record could be incorrectly inserted or updated after the conditional search has been performed

TODO:
- More unit test coverage is required that covers core paths and edge cases

## Related issues
Addresses #236, ab#63056.

## Testing
- All existing tests should pass.
- Adds new Unit tests to validate Handler logic
- Adds new E2E tests with conditional update scenarios

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            IMediator mediator,
+            IsEnabled featureEnabled)
+            : base(fhirDataStore, conformanceProvider, resourceWrapperFactory)
+        {
+            EnsureArg.IsNotNull(searchService, nameof(searchService));
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+
+            _searchService = searchService;
+            _mediator = mediator;
+            _featureEnabled = featureEnabled();
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs]: EnsureArg.IsNotNull(featureEnabled, nameof(featureEnabled));

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests_ConditionalUpsert.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+{
+    /// <summary>
+    /// Tests scenarios for the Conditional Upsert logic
+    /// </summary>
+    public partial class ResourceHandlerTests
+    {
+        [Fact]
+        public async Task WhenUpsertingConditionallyWithNoIdAndNoExisting_GivenAResource_TheServerShouldReturnTheUpdatedResourceSuccessfully()
+        {
+            IReadOnlyList<Tuple<string, string>> list = new[] { Tuple.Create("_tag", Guid.NewGuid().ToString()) };
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests_ConditionalUpsert.cs]: Test if no search criteria

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalUpdates/ConditionalUpsertFeatureModule.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+using Microsoft.Health.Fhir.Api.Configs;
+using Microsoft.Health.Fhir.Core.Features.Resources.Upsert;
+
+namespace Microsoft.Health.Fhir.Api.Modules.FeatureFlags.ConditionalUpdates
+{
+    public class ConditionalUpsertFeatureModule : IStartupModule
+    {
+        public void Load(IServiceCollection services)
+        {
+            services.Add<ConditionalUpsertResourceHandler.IsEnabled>(x => () => x.GetRequiredService<IOptions<FeatureConfiguration>>().Value.SupportsConditionalUpdate)
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalUpdates/ConditionalUpsertFeatureModule.cs]: This is nice, but I wonder if it would be simpler for the constructor to just take a FeatureConfiguration

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalUpdates/ConditionalUpsertFeatureModule.cs]: We could. The next question is does it make sense to move Configuration into Core? This was the original constraint for this

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        public delegate bool IsEnabled();
+
+        public async Task<UpsertResourceResponse> Handle(ConditionalUpsertResourceRequest message, CancellationToken cancellationToken = default)
+        {
+            EnsureArg.IsNotNull(message, nameof(message));
+
+            SearchResult results = await _searchService.SearchAsync(message.Resource.InstanceType, message.ConditionalParameters, cancellationToken);
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs]: From what I understand, a server side javascript in Cosmod DB can act within a single transaction (for a single partition). If we used that for the read/update would it avoid the possibility of the contention issue?

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs]: Stored procedures run on a single partition, the problem here is that the search is cross-partition. In sql we could use a serializable transaction to solve this

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalUpdates/ConditionalUpsertFeatureModule.cs (hunk 3 above)
Comment (johnstairs) [reply src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalUpdates/ConditionalUpsertFeatureModule.cs]: Yeah ok, probably more modular this way.

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests.cs [hunk 5]
```diff
@@ -57,21 +63,25 @@ public ResourceHandlerTests()
...
             observationResource.ReadHistory = false;
             observationResource.UpdateCreate = true;
+            observationResource.ConditionalUpdate = true;
             observationResource.Versioning = CapabilityStatement.ResourceVersionPolicy.Versioned;
 
             CapabilityStatementMock.SetupMockResource(_conformanceStatement, ResourceType.Patient, null);
             var patientResource = _conformanceStatement.Rest.First().Resource.Find(x => x.Type == ResourceType.Patient);
             patientResource.ReadHistory = true;
             patientResource.UpdateCreate = true;
+            observationResource.ConditionalUpdate = true;
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests.cs]: `patientResource.ConditionalUpdate = true` ?

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests_ConditionalUpsert.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+            UpsertResourceResponse result = await _mediator.Send<UpsertResourceResponse>(message);
+
+            Assert.Equal(SaveOutcomeType.Created, result.Outcome.Outcome);
+
+            await _fhirDataStore.Received().UpsertAsync(Arg.Is<ResourceWrapper>(x => x.ResourceId == result.Outcome.Resource.Id), null, true, true, Arg.Any<CancellationToken>());
+        }
+
+        [Fact]
+        public async Task GivenAResource_WhenUpsertingConditionallyWithAnIdAndNoExisting_ThenTheServerShouldReturnTheUpdatedResourceSuccessfully()
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests_ConditionalUpsert.cs]: nit: `ThenTheServerShouldReturnTheCreatedResourceSuccessfully`

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalUpdateTests.cs [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            FhirResponse<Observation> updateResponse = await Client.ConditionalUpdateAsync(
+                observation2,
+                $"identifier={identifier}");
+
+            Assert.Equal(HttpStatusCode.OK, updateResponse.StatusCode);
+
+            Observation updatedResource = updateResponse.Resource;
+
+            Assert.NotNull(updatedResource);
+            Assert.Equal(response.Resource.Id, updatedResource.Id);
```
Comment (namadabu) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalUpdateTests.cs]: Should we also check whether the updated resource has the new Text element? (here and in the previous test)

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalUpdateTests.cs [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+            observation.Identifier.Add(new Identifier("http://e2etests", identifier));
+
+            FhirResponse<Observation> response = await Client.CreateAsync(observation);
+            Assert.Equal(HttpStatusCode.Created, response.StatusCode);
+
+            FhirResponse<Observation> response2 = await Client.CreateAsync(observation);
+            Assert.Equal(HttpStatusCode.Created, response2.StatusCode);
+
+            var observation2 = Samples.GetDefaultObservation().ToPoco<Observation>();
+            observation2.Id = Guid.NewGuid().ToString();
```
Comment (namadabu) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalUpdateTests.cs]: nit: We should set observation2.Id to either `null` or `response.Resource.Id`. Just to make sure we don't hit the IncorrectId flow (which the previous test tests for). The Assertion for the PreconditionFailed does make sure that we are hitting the right scenario, but just thought it would be cleaner to set the Id correctly.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            EnsureArg.IsNotNull(featureEnabled, nameof(featureEnabled));
+
+            _searchService = searchService;
+            _mediator = mediator;
+            _featureEnabled = featureEnabled();
+        }
+
+        public delegate bool IsEnabled();
+
+        public async Task<UpsertResourceResponse> Handle(ConditionalUpsertResourceRequest message, CancellationToken cancellationToken = default)
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs]: Maybe add something about transactions as a todo...

Comment (johnstairs) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs]: Oh, you have that below

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+                }
+            }
+            else if (count == 1)
+            {
+                ResourceWrapper resourceWrapper = results.Results.First();
+                var resource = message.Resource.ToPoco();
+                var version = WeakETag.FromVersionId(resourceWrapper.Version);
+
+                // One Match, no resource id provided OR (resource id provided and it matches the found resource): The server performs the update against the matching resource
+                if (string.IsNullOrEmpty(resource.Id) || resource.Id == resourceWrapper.ResourceId)
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs]: Do an ordinal string comparison

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/XmlFormatter/XmlFormatterConfiguration.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.AspNetCore.Mvc.Formatters;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Api.Configs;
+using Microsoft.Health.Fhir.Api.Features.ContentTypes;
+using Microsoft.Health.Fhir.Core.Features.Conformance;
+
+namespace Microsoft.Health.Fhir.Api.Modules.FeatureFlags.XmlFormatter
+{
+    internal class XmlFormatterConfiguration : IPostConfigureOptions<MvcOptions>, IProvideCapability
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/XmlFormatter/XmlFormatterConfiguration.cs]: So this is not related to actually configuting MvcOptions?
